### PR TITLE
Introduce a manager interface used to add/remove observers, and READM…

### PR DESCRIPTION
**Pull request by objectiser**
…E updated to include spring autoconfigure component

### CODE REVIEW

**Comment by objectiser** [comment]
@sjoerdtalsma Does this address your requirement from https://github.com/opentracing-contrib/java-api-extensions/issues/5#issuecomment-320985909 ?

**Comment by sjoerdtalsma** [comment]
Question: Will you make the current unconditional wrapping conditional too in this PR, or do you want to tackle those in separate PR's? 

(in the spring `TracerBeanPostProcessor` for example)

**Comment by objectiser** [comment]
@sjoerdtalsma Good point - best to handle it in this PR. I'll update it later or tomorrow am.

**Comment by sjoerdtalsma** [comment]
Ok, I'll review after the update :+1:

File: README.md [hunk 1]
```diff
@@ -38,12 +38,13 @@ for each call to `TracerObserver.onStart()`.
 
 ## Registering API extensions
 
-There are two ways an extension API can be registered for use with a `Tracer`.
+There are three ways an extension API can be registered for use with a `Tracer`.
```
Comment (sjoerdtalsma) [inline README.md]: three -> several for less future edits? :wink:

File: README.md [hunk 2]
```diff
@@ -38,12 +38,13 @@ for each call to `TracerObserver.onStart()`.
...
 
 1) Native support within the `Tracer` implementation
 
 Some `Tracer` implementations may decide to implement support for the extension APIs directly, in which
```
Comment (sjoerdtalsma) [inline README.md]: Brevity
>...  , in which case they will implement the `APIExtensionsManager` interface ...

can be simply
>... by implementing the `APIExtensionsManager` interface ...

(Maybe make the APIExtensionsManager a link to the interface itself?)

File: README.md [hunk 3]
```diff
@@ -58,11 +59,30 @@ case an implementation specific mechanism will be provided for registering the A
 The `io.opentracing.contrib.api.tracer.APIExtensionsTracer` provides a single constructor which is supplied
```
Comment (sjoerdtalsma) [inline README.md]: Consistency. Why write the fully-qualified name here and `APIExtensionsManager` above?
I guess it's easier on the eyes to write the short name and maybe link to the java file?

File: README.md [hunk 4]
```diff
@@ -58,11 +59,30 @@ case an implementation specific mechanism will be provided for registering the A
...
+<dependency>
+  <groupId>io.opentracing.contrib</groupId>
+  <artifactId>opentracing-api-extensions-tracer-spring-autoconfigure</artifactId>
+</dependency>
+
+```
+
+Using this approach, `TracerObserver` instances can be defined as Spring `@Bean`s, and they will automatically be
+detected and registered with the API extensions tracer.
+
```
Comment (sjoerdtalsma) [inline README.md]: Maybe change to something like
> Using this approach, any `TracerObserver` Spring `@Bean`s will be automatically detected and registered with the API extensions tracer.

File: opentracing-api-extensions-tracer-spring-autoconfigure/src/main/java/io/opentracing/contrib/api/tracer/spring/autoconfigure/TracerBeanPostProcessor.java [hunk 5]
```diff
@@ -38,21 +39,34 @@
...
-                log.info("Use extensions API tracer to wrap tracer=" + bean + " with observers=" + tracerObservers);
+                log.info("Initialized extensions API manager/tracer="
+                        + tracer + " with observers=" + tracerObservers);
                 return tracer;
             }
         }
         return bean;
     }
 
+    static APIExtensionsManager getManager(Tracer tracer) {
```
Comment (sjoerdtalsma) [inline opentracing-api-extensions-tracer-spring-autoconfigure/src/main/java/io/opentracing/contrib/api/tracer/spring/autoconfigure/TracerBeanPostProcessor.java]: Can be `private static`

File: opentracing-api-extensions-tracer-spring-autoconfigure/src/main/java/io/opentracing/contrib/api/tracer/spring/autoconfigure/TracerBeanPostProcessor.java [hunk 6]
```diff
@@ -38,21 +39,34 @@
...
+        APIExtensionsManager manager = null;
+        if (tracer instanceof APIExtensionsManager) {
+            manager = (APIExtensionsManager)tracer;
+            log.info("Use existing extensions API manager/tracer=" + manager);
+        } else {
+            manager = new APIExtensionsTracer(tracer);
+            log.info("Use extensions API wrapper tracer=" + manager + " to wrap original tracer=" + tracer);
+        }
```
Comment (sjoerdtalsma) [inline opentracing-api-extensions-tracer-spring-autoconfigure/src/main/java/io/opentracing/contrib/api/tracer/spring/autoconfigure/TracerBeanPostProcessor.java]: Personally I wouldn't use `log.info` here, but rather `debug` or even lower.
There is already an info message in the main method `"Initialized extensions API...."`.

However, as this is only used during Spring initialization, I don't feel strongly about it, just my opinion.

File: opentracing-api-extensions-tracer-spring-autoconfigure/src/test/java/io/opentracing/contrib/api/tracer/spring/autoconfigure/TracerBeanPostProcessorTest.java [hunk 7]
```diff
@@ -73,4 +77,17 @@ public void testTracerWrapped() {
         Mockito.verify(tracerObserver).onStart(Matchers.any(SpanData.class));
     }
 
+    @Test
+    public void testGetWrappedTracerManager() {
+        assertTrue(TracerBeanPostProcessor.getManager(mockTracer) instanceof APIExtensionsTracer);
+    }
+
```
Comment (sjoerdtalsma) [inline opentracing-api-extensions-tracer-spring-autoconfigure/src/test/java/io/opentracing/contrib/api/tracer/spring/autoconfigure/TracerBeanPostProcessorTest.java]: I get a feeling that you're testing the inner workings here.
What you care about should be the postprocess functionality right?

Now I understand why you've opened up the internal method.

**Review by sjoerdtalsma — COMMENTED** [review]
All in all looks good, just some minor comments.

This interface indeed addresses my concern and could enable additional postprocessors each adding a single concern while avoiding re-wrapping the tracer each time.

**Comment by objectiser** [comment]
@sjoerdtalsma Thanks for the comments - hopefully all fixed.

**Review by sjoerdtalsma — APPROVED** [review]
Yup, all fixed indeed, lgtm :+1:
Thanks!

**Pull request merged**